drivers/at86rf2xx/aes_spi.c

```diff
@@ -0,0 +1,432 @@
+                                    uint8_t key[AT86RF2XX_AES_KEY_LENGTH])
+{
+    uint8_t cfg = AT86RF2XX_AES_CTRL_AES_MODE__KEY |
+                  AT86RF2XX_AES_CTRL_AES_DIR__ENC;
+    at86rf2xx_spi_get_bus(dev);
+    _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+    _aes_transfer_bytes(dev, false, &cfg, NULL, sizeof(cfg));
+    _aes_open_read(dev, AT86RF2XX_REG__AES_KEY_START);
+    _aes_transfer_bytes(dev, false, NULL, key, AT86RF2XX_AES_KEY_LENGTH);
+    at86rf2xx_spi_release_bus(dev);
+}
+
+void at86rf2xx_aes_key_write_encrypt(at86rf2xx_t *dev,
+                                     const uint8_t key[AT86RF2XX_AES_KEY_LENGTH])
+{
+    uint8_t cfg = AT86RF2XX_AES_CTRL_AES_MODE__KEY |
+                  AT86RF2XX_AES_CTRL_AES_DIR__ENC;
+    at86rf2xx_spi_get_bus(dev);
+    _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+    _aes_transfer_bytes(dev, true, &cfg, NULL, sizeof(cfg));
+    _aes_transfer_bytes(dev, false, key, NULL, AT86RF2XX_AES_KEY_LENGTH);
+    at86rf2xx_spi_release_bus(dev);
+}
+
+void at86rf2xx_aes_key_read_decrypt(at86rf2xx_t *dev,
+                                    uint8_t key[AT86RF2XX_AES_KEY_LENGTH])
+{
+    uint8_t cfg = AT86RF2XX_AES_CTRL_AES_MODE__KEY |
+                  AT86RF2XX_AES_CTRL_AES_DIR__DEC;
+    at86rf2xx_spi_get_bus(dev);
+    _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+    _aes_transfer_bytes(dev, false, &cfg, NULL, sizeof(cfg));
+    _aes_open_read(dev, AT86RF2XX_REG__AES_KEY_START);
+    _aes_transfer_bytes(dev, false, NULL, key, AT86RF2XX_AES_KEY_LENGTH);
+    at86rf2xx_spi_release_bus(dev);
+}
+
+void at86rf2xx_aes_key_write_decrypt(at86rf2xx_t *dev,
+                                     const uint8_t key[AT86RF2XX_AES_KEY_LENGTH])
+{
+    uint8_t cfg = AT86RF2XX_AES_CTRL_AES_MODE__KEY |
+                  AT86RF2XX_AES_CTRL_AES_DIR__DEC;
+    at86rf2xx_spi_get_bus(dev);
+    _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+    _aes_transfer_bytes(dev, true, &cfg, NULL, sizeof(cfg));
+    _aes_transfer_bytes(dev, false, key, NULL, AT86RF2XX_AES_KEY_LENGTH);
+    at86rf2xx_spi_release_bus(dev);
+}
+
+int at86rf2xx_aes_ecb_encrypt(at86rf2xx_t *dev,
+                              aes_block_ptr_t cipher,
+                              uint8_t key[AT86RF2XX_AES_BLOCK_SIZE],
+                              const aes_block_ptr_t plain,
+                              uint8_t nblocks)
+{
+    uint8_t status = 0;
+    if (nblocks) {
+        uint8_t cfg = AT86RF2XX_AES_CTRL_AES_MODE__ECB |
+                      AT86RF2XX_AES_CTRL_AES_DIR__ENC;
+        uint8_t mirror = cfg | AT86RF2XX_AES_CTRL_MIRROR_MASK__AES_REQUEST;
+        at86rf2xx_spi_get_bus(dev);
+        _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+        _aes_transfer_bytes(dev, true, &cfg, NULL, sizeof(cfg));
+        _aes_transfer_bytes(dev, true, plain[0], NULL, AT86RF2XX_AES_BLOCK_SIZE);
+        _aes_transfer_bytes(dev, false, &mirror, NULL, sizeof(mirror));
+
+        xtimer_usleep(AT86RF2XX_AES_DELAY_US);
```

Comment:
Do we get a bit in the interrupt status register set when the encryption is done?

I'd like to avoid having to rely entirely on a delay - maybe the clocks on the MCU are not accurate.
And increasing the delay to accommodate for that would just hurt performance, so I'd rather have a tight delay and then spin on reading a status register until the 'done' bit is set.

Reply:
Ah, never mind - that's exactly what `_aes_check()` is doing.

Reply:
OK, then I´d rename the combination of both to `_aes_wait_for_result()`.